### ISSUE TITLE
There are no double disambiguation bishop capture checkmates, but there could have been

### ISSUE THREAD

**Issue by maxbergmark** [paralogical/rarest-move-in-chess#3]
First of all, great work with the video! I asked myself the very same question you did two years ago, but back then I decided to avoid looking at disambiguations to make it simpler. But after seeing your video, I realized that the code I wrote could be used to find some kind of answer to your question. Here's [the repo](https://github.com/maxbergmark/rust-chess-stat-recorder) for my project. 

Much like you, I parsed all games in the lichess database, but I was mainly trying to find the probability that a player missed a win (mate in 1). In order to figure this out, I not only needed to check which moves were played, but if there was another move in the position that would have been a checkmate.

You might want to have a look at [this game](https://lichess.org/WFd7yYEP#117). While there wasn't a double disambiguation capture bishop checkmate in there, black could have chosen to play Ba3xb2# if they wanted to. While this isn't the definitive answer to which move is the most rare, there is at least some consolidation knowing that it could have been played at some point.

![image](https://github.com/paralogical/rarest-move-in-chess/assets/1706486/81d05d70-9b69-484c-acff-e7db197c293d)

So far, I have only parsed games from 2013-2020 since rewriting the parser, but in those games I also managed to find a [missed knight double disambiguation checkmate](https://lichess.org/q38Hhq1P#108). I'll see if I can go through the rest of the games soon, and find some more examples.

**Comment by paralogical** [comment]
Wow, this is super cool! Thanks for sharing!

**Comment by maxbergmark** [comment]
I have looked through all the games now, and I added support for finding some more types of rare checkmates. For each checkmate, I recorded whether it was actually played, or just a move that could have been played in any position. The two categories of checkmates that I looked at were king checkmates and double disambiguation checkmates.

With checking move variation, the analysis included a total of 11.17 trillion moves, which takes around 12 hours to go through on my machine. 

For double disambiguation checkmates, I further categorized them by whether or not they included captures, and which type of checkmate they were. The three types of checkmates in this context are `Normal`, `Discovered`, and `Double`. By contrast, any king checkmate is a discovered checkmate by definition. 

I don't have a formal proof for it, but I'm fairly sure that bishop double disambiguation checkmates can't be double checkmates. It would require the enemy king to be on the same square as one of the other bishops leading to the disambiguation from what I can tell. But I'm happy to be proven wrong.
### Actually played mates
These should align with your results, but are using data up until June 2024.
- Doubly disambiguated knight checkmates:
	- Normal: 48
	- Discovered: 3
	- Double: 0
- Doubly disambiguated bishop checkmates:
	- Normal: 0
	- Discovered: 2 (one played since your video)
- Doubly disambiguated knight capture checkmates:
	- Normal: 1 (played since your video)
	- Discovered: 3 (all played since your video, one was a bullet game)
	- Double: 0
- Doubly disambiguated bishop capture checkmates:
	- Normal: 1 (played since your video)
	- Discovered: 1 (played since your video)

### Potential mates
- Doubly disambiguated knight checkmates:
	- Normal: 220
	- Discovered: 35
	- Double: 3
- Doubly disambiguated bishop checkmates:
	- Normal: 0
	- Discovered: 14 (half of them from [this game](https://lichess.org/zozCPC2w#133))
- Doubly disambiguated knight capture checkmates:
	- Normal: 2
	- Discovered: 5
	- Double: 0
- Doubly disambiguated bishop capture checkmates:
	- Normal: 2
	- Discovered: 0

Unfortunately, I don't think this data reveals any direct conclusion as to which checkmate is more rare. My bet would be on Doubly disambiguated knight capture double checkmates or Doubly disambiguated bishop capture normal checkmates. But it will be difficult to get unbiased data from now on, since lots of people watched your video and will try to achieve the rarest checkmate. I don't think that's a bad thing necessarily, so don't take it as criticism. 

## Some notable games
- [Potential double disambiguation discovered knight checkmate](https://lichess.org/mYSthdbG#104)
- [Potential double disambiguation discovered knight double checkmate](https://lichess.org/1GfhDOBO#170)
- [Another potential double disambiguation discovered knight double checkmate](https://lichess.org/VZT5TsS6#121)
- [Potential double disambiguation capture bishop checkmate](https://lichess.org/WFd7yYEP)
- [Potential double disambiguation capture knight checkmate](https://lichess.org/q38Hhq1P)
- [Double disambiguation bishop checkmate played after video](https://lichess.org/UcZZx10k#99)

## Output format
To be able to check the output and count different types of mates, I just wrote all rare checkmates to a CSV file. Each row contains the move number, the move SAN, an encoding of the checkmate type, and a link to the game.

```
 30.   O-O-O# ,K   ?,https://lichess.org/67jlylfa
 79... Kg6#   ,K   ?,https://lichess.org/E9uo1ERz
106.   Nd4f3# ,DD  ?,https://lichess.org/TjuaWLcF
 64... Kc7#   ,K    ,https://lichess.org/Y60sstLS
 64... Kb7#   ,K   ?,https://lichess.org/Y60sstLS
...
```
The checkmate type is encoded as follows:
- King (`K `) or double disambiguation (`DD`)
- Capture or not (`x` or ` `)
- Checkmate type for double disambiguation (` `, `D` or `2` for normal, discovered, and double)
- Was it actually played? (` ` if played, `?` if potential move)

This way, I can use `cat output/*.moves | grep -P ",DDxD "` to find all games with double disambiguation capture discovered checkmates. <s>I'll see if I can share this dataset in my repo. Right now it's a bit too large, since it includes a lot of king checkmates. But most of those aren't that rare, so they can be excluded from the data.</s> Here are [all rare checkmates that I found](https://github.com/maxbergmark/rust-chess-stat-recorder/blob/main/rare_checkmates.csv). 

## Non-disambiguated checkmates
As a side-note, I noticed that you mentioned the king capture mate Kxa1# in your video. I'm not sure if it was intentional or not, but there are actually just two non-disambiguated moves that have never been played on lichess, and those are Kxa1# and Kxa8#.
